### PR TITLE
Operator can add and remove ES nodes

### PR DESCRIPTION
**Pull request by josefkarasek**
Scaling up:
* update config map with new variables based on data and master nodes
* add new nodes to the cluster
* refresh configuration of running nodes, that don't catch the changes from the config map

Scaling down:
* change config map
* refresh configuration of running nodes
* remove nodes

No data guarantees for data persistency when scaling down!!

Note: we need to carefully set the quorum variable so that we don't accidentally create a dead-lock between master nodes after scaling up and before scaling down.

Why after scaling up:
* If we changed it before, the expected quorum would be greater that the number of Ready master nodes

Why before scaling down:
* If we changed it after, the cluster would expect greater number of Ready master nodes, while nodes to-be-removed nodes leave the cluster.

### CODE REVIEW

File: pkg/k8shandler/cluster.go [hunk 1]
```diff
@@ -47,22 +48,22 @@ func CreateOrUpdateElasticsearchCluster(dpl *v1alpha1.Elasticsearch, configMapNa
 		return err
 	}
 	// TODO: get rid of this message as part of LOG-206
-	logrus.Infof("cluster %s required action is: %v", dpl.Name, action)
+	// logrus.Infof("cluster %s required action is: %v", dpl.Name, action)
```
Comment (richm) [inline pkg/k8shandler/cluster.go]: Is it worth keeping this and making it a `Debugf`?

File: pkg/k8shandler/cluster.go [hunk 2]
```diff
@@ -274,3 +306,89 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
...
 	return nil
 }
+
+func updateClusterSettings(dpl *v1alpha1.Elasticsearch) error {
+
+	event, err := getClusterEvent(dpl)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Event is: %s\n", event)
```
Comment (richm) [inline pkg/k8shandler/cluster.go]: `logrus.Debugf` instead?

File: pkg/k8shandler/cluster.go [hunk 3]
```diff
@@ -274,3 +306,89 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
...
+	_, scaleDownCondition := utils.GetESNodeCondition(&dpl.Status, v1alpha1.ScalingDown)
+	_, scaleUpCondition := utils.GetESNodeCondition(&dpl.Status, v1alpha1.ScalingUp)
+
+	if settingsUpdateCondition != nil && settingsUpdateCondition.Status == v1alpha1.ConditionTrue {
+		if scaleDownCondition != nil && scaleDownCondition.Status == v1alpha1.ConditionTrue {
+			// TODO: here we can wait till no pods are Terminating
+			return v1alpha1.ScaledDown, nil
+		}
+		if scaleUpCondition != nil && scaleUpCondition.Status == v1alpha1.ConditionTrue {
+			pods, err := listPods(dpl.Name, dpl.Namespace)
```
Comment (richm) [inline pkg/k8shandler/cluster.go]: do you somehow need to wait for all expected pods?  e.g. what if `listPods` returns only 3 pods but because of the scaleup now 5 are expected?

File: pkg/k8shandler/configuration_tmpl.go [hunk 4]
```diff
@@ -19,11 +19,11 @@ network:
 
 discovery.zen:
   ping.unicast.hosts: {{.EsUnicastHost}}
-  minimum_master_nodes: ${NODE_QUORUM}
+  minimum_master_nodes: {{.NodeQuorum}}
```
Comment (richm) [inline pkg/k8shandler/configuration_tmpl.go]: how was `${NODE_QUORUM}` evaluated previously?  As an environment variable at runtime?  Which means the user could change the value by changing the deployment env var?

File: pkg/k8shandler/deployment.go [hunk 5]
```diff
@@ -54,6 +54,7 @@ func (node *deploymentNode) isDifferent(cfg *desiredNodeState) (bool, error) {
 		if value.ValueFrom == nil {
 			if desiredVars[index] != value {
 				logrus.Infof("Env vars are different for %q", node.resource.GetName())
+				fmt.Printf("%s=%s cahnged to %s=%s\n", value.Name, value.Value, desiredVars[index].Name, desiredVars[index].Value)
```
Comment (richm) [inline pkg/k8shandler/deployment.go]: should this be `Debugf` instead?  `changed` instead of `cahnged`

File: pkg/k8shandler/desirednodestate.go [hunk 6]
```diff
@@ -342,14 +342,6 @@ func (cfg *desiredNodeState) getEnvVars() []v1.EnvVar {
...
 			Value: heapDumpLocation,
 		},
-		v1.EnvVar{
-			Name:  "NODE_QUORUM",
-			Value: strconv.Itoa(int(cfg.MasterNum/2 + 1)),
-		},
-		v1.EnvVar{
-			Name:  "RECOVER_EXPECTED_NODES",
-			Value: strconv.Itoa(int(cfg.DataNum)),
-		},
```
Comment (richm) [inline pkg/k8shandler/desirednodestate.go]: so users will need to edit the configmap instead of being able to set these as env. vars. in the deployment?

Comment (richm) [reply pkg/k8shandler/desirednodestate.go]: Maybe that's better since you can set it in one place with a configmap vs. each deployment (until we move to statefulset . . .)

File: pkg/k8shandler/status.go [hunk 7]
```diff
@@ -98,7 +109,13 @@ func clusterHealth(dpl *v1alpha1.Elasticsearch) string {
 		return healthUnknown
 	}
 
-	return result["status"].(string)
+	fmt.Printf("received: '%v'\n", result["status"])
```
Comment (richm) [inline pkg/k8shandler/status.go]: `logrus.Debugf`?

File: pkg/stub/handler.go [hunk 8]
```diff
@@ -34,7 +33,7 @@ func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
 // Reconcile reconciles the cluster's state to the spec specified
 func Reconcile(es *v1alpha1.Elasticsearch) (err error) {
 	// TODO: get rid of this message as part of LOG-206
-	logrus.Info("Started reconciliation")
+	// logrus.Info("Started reconciliation")
```
Comment (richm) [inline pkg/stub/handler.go]: `logrus.Debug`?

File: pkg/k8shandler/configuration_tmpl.go (hunk 4 above)
Comment (josefkarasek) [reply pkg/k8shandler/configuration_tmpl.go]: Yes, it was an environment variable. We're setting the quorum dynamically now based on the number of master nodes in the cluster: `(master_eligible_nodes / 2) + 1`

https://www.elastic.co/guide/en/elasticsearch/reference/6.4/discovery-settings.html#minimum_master_nodes

File: pkg/k8shandler/desirednodestate.go (hunk 6 above)
Comment (josefkarasek) [reply pkg/k8shandler/desirednodestate.go]: We don't allow modification of the configuration stored in the config map.
We run sha256 on the config map content and revert all changes made by 'someone else other than the operator'

What we do allow users to change though is `REPLICA_SHARDS`. And that's only because we aren't able to deliver `resiliency flavours` (no replication, some replication, full replication) in time for the release. @ewolinetz can say more about this.

File: pkg/apis/elasticsearch/v1alpha1/types.go [hunk 9]
```diff
@@ -121,3 +121,44 @@ const (
...
+
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+type ClusterEvent string
```
Comment (ewolinetz) [inline pkg/apis/elasticsearch/v1alpha1/types.go]: This type seems unused

File: pkg/k8shandler/cluster.go [hunk 10]
```diff
@@ -274,3 +306,89 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
 	}
 	return nil
 }
+
+func updateClusterSettings(dpl *v1alpha1.Elasticsearch) error {
+
```
[the remark this answers is not in the document]
Comment (ewolinetz) [reply pkg/k8shandler/cluster.go]: I'm not sure... on one hand it makes sense because its part of `cluster` logic. I think for now lets keep it in here and we can evaluate doing a refactor later since I feel like a lot of logic in the operator can be better organized.

File: pkg/k8shandler/cluster.go [hunk 11]
```diff
@@ -274,3 +306,89 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
...
+	fmt.Printf("Event is: %s\n", event)
+
+	switch event {
+	case v1alpha1.ScaledDown:
+		if err := execClusterSettingsUpdate(dpl); err != nil {
+			return err
+		}
+		utils.UpdateClusterSettingsUpdateNeededCondition(&dpl.Status, v1alpha1.ConditionFalse)
+		utils.UpdateScalingDownCondition(&dpl.Status, v1alpha1.ConditionFalse)
+		break
```
Comment (ewolinetz) [inline pkg/k8shandler/cluster.go]: I don't think we actually need to do an explicit `break` for `switch` statements in go

File: pkg/k8shandler/cluster.go [hunk 12]
```diff
@@ -274,3 +306,89 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
...
+
+func updateClusterSettings(dpl *v1alpha1.Elasticsearch) error {
+
+	event, err := getClusterEvent(dpl)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Event is: %s\n", event)
+
+	switch event {
```
Comment (ewolinetz) [inline pkg/k8shandler/cluster.go]: I'm wondering if this should be a generic `switch { }` statement and instead of getting a particular event, we check the cluster conditions values and rely on them being updated prior to getting here.

File: pkg/k8shandler/desirednodestate.go (hunk 6 above)
Comment (ewolinetz) [reply pkg/k8shandler/desirednodestate.go]: I don't see why we couldn't incorporate the `REPLICA_SHARDS` change as part of this.

File: pkg/utils/exec/exec.go [hunk 13]
```diff
@@ -1,4 +1,4 @@
-package k8shandler
+package exec
```
Comment (ewolinetz) [inline pkg/utils/exec/exec.go]: we could just keep this as `pkg/utils/exec.go` then it would be package `utils` and we would get it when we do an import of `utils`.
Unless you have a specific reasoning for making it separate?

File: pkg/k8shandler/cluster.go (hunk 3 above)
Comment (josefkarasek) [reply pkg/k8shandler/cluster.go]: @richm I changed it so that the number of currently running nodes is read from `es-logging:9200/_cluster/health["number_of_nodes"]`

What do you think?

File: pkg/k8shandler/cluster.go (hunk 1 above)
Comment (josefkarasek) [reply pkg/k8shandler/cluster.go]: log-206 merged

File: pkg/k8shandler/cluster.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (josefkarasek) [reply pkg/k8shandler/cluster.go]: ok, I'll keep it where it is now

File: pkg/k8shandler/cluster.go (hunk 2 above)
Comment (josefkarasek) [reply pkg/k8shandler/cluster.go]: log-206 merged

File: pkg/k8shandler/status.go (hunk 7 above)
Comment (josefkarasek) [reply pkg/k8shandler/status.go]: log-206 merged

File: pkg/utils/exec/exec.go (hunk 13 above)
Comment (josefkarasek) [reply pkg/utils/exec/exec.go]: moved

File: pkg/k8shandler/cluster.go (hunk 3 above)
Comment (richm) [reply pkg/k8shandler/cluster.go]: > @richm I changed it so that the number of currently running nodes is read from es-logging:9200/_cluster/health["number_of_nodes"]

That could still return less than the expected number of nodes

File: pkg/apis/elasticsearch/v1alpha1/types.go [hunk 14]
```diff
@@ -136,3 +136,45 @@ const (
...
+	// Human-readable message indicating details about last transition.
+	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
+}
+
+// ClusterConditionType is a valid value for ClusterCondition.Type
+type ClusterConditionType string
+
+// These are valid conditions for elasticsearch node
+const (
+	ClusterSettingsUpdateNeeded ClusterConditionType = "SettingsUpdateNeeded"
```
[the remark this answers is not in the document]
Comment (ewolinetz) [reply pkg/apis/elasticsearch/v1alpha1/types.go]: This is for the operator to use to know what it should be doing. This is different than `scaling up`, which has a definition below.

File: pkg/k8shandler/cluster.go (hunk 3 above)
Comment (josefkarasek) [reply pkg/k8shandler/cluster.go]: Yes, we wait for all nodes to join the cluster before we mark the scaleUp operation as complete.
Similarly for scaleDown, except those nodes leave the cluster.

https://github.com/openshift/elasticsearch-operator/pull/33/files#diff-b0f11234653f638f62ef2fc4f695a69dR380

File: pkg/k8shandler/desirednodestate.go (hunk 6 above)
Comment (josefkarasek) [reply pkg/k8shandler/desirednodestate.go]: openshift/origin-aggregated-logging/pull/1468

File: pkg/apis/elasticsearch/v1alpha1/types.go (hunk 9 above)
Comment (josefkarasek) [reply pkg/apis/elasticsearch/v1alpha1/types.go]: Values of this type mark certain events that happen in the cluster
Eg node quorum is update only if we know know that the cluster scaled-up successfully - `ScaledUp` event.
```go
const (
	ScaledDown            ClusterEvent = "ScaledDown"
	ScaledUp              ClusterEvent = "ScaledUp"
	UpdateClusterSettings ClusterEvent = "UpdateClusterSettings"
	NoEvent               ClusterEvent = "NoEvent"
)
```

File: pkg/k8shandler/cluster.go (hunk 12 above)
Comment (josefkarasek) [reply pkg/k8shandler/cluster.go]: That's exactly what `getClusterEvent` does.
To the code makes more sense when this is divided into two separate functions:
1) based on conditions figure out what is happening in the cluster
2) act based on the above information

File: pkg/apis/elasticsearch/v1alpha1/types.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (josefkarasek) [reply pkg/apis/elasticsearch/v1alpha1/types.go]: @jcantrill we dynamically change cluster settings, in this example node quorum, based on the number of nodes in the cluster

[the remark this answers is not in the document]
Comment (jcantrill) [reply pkg/apis/elasticsearch/v1alpha1/types.go]: My point is I think it reads "I need you to do something" instead of saying "I am doing soming".  I believe its a subtle choice in wording that should be corrected

**Comment by josefkarasek** [comment]
openshift/origin-aggregated-logging/pull/1477

File: pkg/apis/elasticsearch/v1alpha1/types.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ewolinetz) [reply pkg/apis/elasticsearch/v1alpha1/types.go]: I'd suggest `UpdatingSettings` then

[the remark this answers is not in the document]
Comment (josefkarasek) [reply pkg/apis/elasticsearch/v1alpha1/types.go]: changed to UpdatingSettings

File: pkg/k8shandler/desirednodestate.go (hunk 6 above)
Comment (josefkarasek) [reply pkg/k8shandler/desirednodestate.go]: I put REPLICA_SHARDS in the same config map as PRIMARY_SHARDS

File: pkg/apis/elasticsearch/v1alpha1/types.go [hunk 15]
```diff
@@ -30,13 +30,21 @@ type Elasticsearch struct {
...
+
 // ElasticsearchSpec struct represents the Spec of Elasticsearch cluster CRD
 type ElasticsearchSpec struct {
 	// managementState indicates whether and how the operator should manage the component
-	ManagementState ManagementState `json:"managementState"`
-
-	Nodes []ElasticsearchNode   `json:"nodes"`
-	Spec  ElasticsearchNodeSpec `json:"nodeSpec"`
+	ManagementState ManagementState       `json:"managementState"`
+	DataReplication DataReplicationType   `json:"dataReplication"`
```
Comment (ewolinetz) [inline pkg/apis/elasticsearch/v1alpha1/types.go]: Maybe rename to something like `ReplicationPolicy` for the spec name

File: pkg/k8shandler/defaults.go [hunk 16]
```diff
@@ -20,10 +22,23 @@ func KibanaIndexMode(mode string) (string, error) {
...
+func rootLogger() string {
 	return "rolling"
 }
+
+func replicaCount(dpl *v1alpha1.Elasticsearch) int {
+	dataNodeCount := int((getDataCount(dpl)))
+	repType := dpl.Spec.DataReplication
+	switch {
+	case repType == v1alpha1.PartialReplication:
+		return (dataNodeCount - 1) / 2
```
Comment (ewolinetz) [inline pkg/k8shandler/defaults.go]: we should check that dataNodeCount > 0

File: pkg/k8shandler/deployment.go [hunk 17]
```diff
@@ -32,7 +32,7 @@ func (node *deploymentNode) isDifferent(cfg *desiredNodeState) (bool, error) {
 	for _, container := range node.resource.Spec.Template.Spec.Containers {
 		if container.Name == "elasticsearch" {
 			if container.Image != cfg.ESNodeSpec.Spec.Image {
-				logrus.Debugf("Container image '%v' is different that desired, updating..", container.Image)
+				logrus.Infof("Container image '%v' for node '%s' is different that desired, updating..", container.Image, node.resource.Name)
```
Comment (ewolinetz) [inline pkg/k8shandler/deployment.go]: We only have one container now. 
What about `("Elasticsearch image change found, updating %q", node.resource.Name)`

File: pkg/k8shandler/util_test.go [hunk 18]
```diff
@@ -2,7 +2,6 @@ package k8shandler
```
Comment (ewolinetz) [inline pkg/k8shandler/util_test.go]: is this file really necessary anymore?

File: pkg/utils/utils.go [hunk 19]
```diff
@@ -40,3 +44,163 @@ func LookupEnvWithDefault(envName, defaultValue string) string {
...
+		// We are adding new node condition.
+		status.Conditions = append(status.Conditions, *condition)
+		return true
+	}
+	// We are updating an existing condition, so we need to check if it has changed.
+	if condition.Status == oldCondition.Status {
+		condition.LastTransitionTime = oldCondition.LastTransitionTime
+	}
+
+	isEqual := condition.Status == oldCondition.Status &&
```
Comment (ewolinetz) [inline pkg/utils/utils.go]: Can we create a function to do this?
```
isEqual := condition.isEqual(oldCondition)

func (condition *api.ClusterCondition)isEqual(oldCondition *api.ClusterCondition) bool {
  return condition.Status == oldCondition.Status && ....
}
```

**Comment by ewolinetz** [comment]
We have a mix of `fmt.Sprintf` and `logrus.{log_level}f` should we pick one over the other for the sake of consistency?

File: pkg/k8shandler/deployment.go (hunk 17 above)
Comment (josefkarasek) [reply pkg/k8shandler/deployment.go]: so...
`"Container image '%v' is different that desired"`

File: pkg/utils/utils.go (hunk 19 above)
Comment (ewolinetz) [reply pkg/utils/utils.go]: I was just thinking to keep this looking neater

File: pkg/k8shandler/cluster.go [hunk 20]
```diff
@@ -274,3 +306,72 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
...
+		}
+		err = utils.UpdateConditionWithRetry(dpl, v1alpha1.ConditionFalse, utils.UpdateUpdatingSettingsCondition)
+		err = utils.UpdateConditionWithRetry(dpl, v1alpha1.ConditionFalse, utils.UpdateScalingUpCondition)
+	case v1alpha1.NoEvent:
+		return nil
+	}
+	return err
+}
+
+func getClusterEvent(dpl *v1alpha1.Elasticsearch, pod *v1.Pod) (v1alpha1.ClusterEvent, error) {
```
Comment (richm) [inline pkg/k8shandler/cluster.go]: this function returns an `error` but no errors are handled and only `nil` is returned

File: pkg/k8shandler/configmaps.go [hunk 21]
```diff
@@ -50,27 +68,49 @@ func createOrUpdateConfigMap(configMapName, namespace, clusterName, kibanaIndexM
...
+			if err := utils.UpdateConditionWithRetry(dpl, v1alpha1.ConditionTrue, utils.UpdateUpdatingSettingsCondition); err != nil {
+				return fmt.Errorf("Unable to update Elasticsearch cluster status: %v", err)
+			}
+
+			return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				if getErr := sdk.Get(existingCM); getErr != nil {
+					logrus.Debugf("Could not get Elasticsearch %v: %v", dpl.Name, getErr)
+					return getErr
+				}
+				if updateErr := sdk.Update(elasticsearchCM); updateErr != nil {
```
Comment (richm) [inline pkg/k8shandler/configmaps.go]: The theory of using `RetryOnConflict` is that the object returned from `Get` may have changed, and you want to preserve the existing fields in that object, and just change the fields you want to change:
```
Get(object)
object.field = value
Update(object)
```
I think if you `Update(elasticsearchCM)` you may be overwriting fields in the CM whose current state is represented by `existingCM`

File: pkg/k8shandler/deployment.go [hunk 22]
```diff
@@ -32,7 +32,7 @@ func (node *deploymentNode) isDifferent(cfg *desiredNodeState) (bool, error) {
 	for _, container := range node.resource.Spec.Template.Spec.Containers {
 		if container.Name == "elasticsearch" {
 			if container.Image != cfg.ESNodeSpec.Spec.Image {
-				logrus.Debugf("Container image '%v' is different that desired, updating..", container.Image)
+				logrus.Debugf("Container image '%v' is different that desired", container.Image)
```
Comment (ewolinetz) [inline pkg/k8shandler/deployment.go]: To me, it would be more useful to know what node/deployment needs to be updated rather than which image

File: pkg/k8shandler/cluster.go [hunk 23]
```diff
@@ -48,18 +51,20 @@ func CreateOrUpdateElasticsearchCluster(dpl *v1alpha1.Elasticsearch, configMapNa
...
 		if err != nil {
 			return err
 		}
 	case action == v1alpha1.ElasticsearchActionScaleDownNeeded:
-		err = cState.removeStaleNodes()
-		if err != nil {
-			return err
-		}
+		// TODO: provide documentation for manual scale down
+		return fmt.Errorf("Scale down operation requested but is not supported by the operator. For manual scale down, follow this document %s", "")
```
Comment (ewolinetz) [inline pkg/k8shandler/cluster.go]: This will keep looping, `TODO` to update that (non-blocker)

**Comment by ewolinetz** [comment]
We will need to update the `CLO` dep and also update the `OLM` spec descriptions
cc @jcantrill
